Comments M. Schuckers 9 Sept 2025
 - Again I love these data.
 - Worried about timeliness of predicting 2025 results mid season.  I get the desire at the moment but not sure it will age well.  Let's update the data from the 2025 reason to get the full season results. - Done
 - Need to more explicitly explain what is happening with scaling and why we are doing it.  Also, we use fit_transform() on training data and tranform() on test data.  Here's why
[< https://towardsdatascience.com/what-and-why-behind-fit-transform-vs-transform-in-scikit-learn-78f915cf96fe/>] - Done. 
 - Maybe go through the 19-23 train with 24 test as an example then have 19-24 train and 25 test as an exercise.
 - Need to have some explanation of the statistics.  What is a field goal?  What is a two-point versus a three-point field goal?  (All field goals in football are 3 points)  What is a rebound?  An assist? etc. - Done.
 - When you do 'log_predictions' explain what a zero represents and what a one represents. I would also rename that variable 'logit_predictions' to make it clearer - Done.
 - logistic regression assumes a linear relationship among the log-odds of the response and the predictors - Done.
 - Comment on the numbers that go into the first accuracy score. Also note that this is training data accuracy. - Done.
 - Be clear and explicit about the results and what they mean. - Done.
 - I've rounded the predicted probabilities for some of the output to help make sense of those values. Probably should mention early on that the probabilities in each row will sum to 1. And explain why that is. - Thank you for doing that, I have added a note about sumation of 1 for each row.


Comments M. Schuckers 11 Aug 2025
- Let's break this first markdown into a couple of cells _- Done._
- Make sure you mention kNN methods earlier _- Done._
- I think there are only 12 rows in these data and that is likely to lead to some overfitting. Your description says 60 rows which would be 5 years of data, I think. _- Done; I compiled 5 years of data for training the models._
- For transforming and rescaling data, you should use the scaler for the training data and apply that same scaler to the test data. This avoids some possible information leakage from the training to the test data. _- Done._ 
- When I run the logistic I don't get any 1's. Just 0's and the probability for the first team is 0.38 of being a one. _- Done. For some reason, when creating the module it gave me a 1 but I just re-ran it and got what you did._

- Also have a categorical response that is only 8% of rows can lead to some issues. Can we make the target, made the playoffs instead? The model can be 92% accurate by just saying all of the rows will be zero. _-Agreed; I have made a seperate notebook file to do this suggestion. The file is located in this module's file on the Github. The title starts with "PLAYOFF VERSION". I'm doing this just so we can keep this version as a backup for now._



---
Title: "Intro to Categorical Model-Making Module by Austin Hayes"

Author:
  - Name: Austin Hayes

  -  Email: ahayes65@charlotte.edu

  -  Affiliation: University of North Carolina at Charlotte

Date: July 29, 2025

Description: Using WNBA Team 'Per Game' stats, we will review the basics of categorical model-making. We will make QDA, LDA, kNN, and Logistic models in order to predict which teams will make the WNBA playoffs this season.

Categories:
  - Logistic Regression
  - Quadratic Discriminant Analysis
  - Linear Discriminant Analysis
  - K Nearest Neighbors
  - Importing and Reading data
  - Data Cleaning
  - Data Science
  - Pandas

### Data

This Dataset is from Basketball Reference @ https://www.basketball-reference.com

Visit the original data page here: https://www.basketball-reference.com/wnba/years/2024.html

The data set contains 60 rows and 26 columns. Each row represents a WNBA team during the 2019-2023 seasons. We also used other datasets including the following seasons: 2020-2024, 2024 ONLY and 2025 ONLY. No matter the dataset, all rows represent a team and the variables stay the same.

Download data: 

Available on the [Intro to Categorical Model-Making Module by Austin Hayes](https://github.com/schuckers/Charlotte_SCORE_Summer25/tree/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes): [2024_WNBA_Per_Game.csv](https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes/2024_WNBA_Per_Game.csv)

---

### Variables and their Descriptions:


<details>
<summary><b>Variable Descriptions</b></summary>

| Variable | Description |
|----------|-------------|
| Rk       | Rank of the team in the league |
| Team     | Name of the team |
| G        | Games played |
| MP       | Minutes played per game |
| FG       | Field Goals made per game |
| FGA      | Field Goal attempts per game |
| FG%      | Field Goal percentage (FG ÷ FGA) |
| 3P       | Three-Point Field Goals made per game |
| 3PA      | Three-Point Field Goal attempts per game |
| 3P%      | Three-Point Field Goal percentage (3P ÷ 3PA) |
| 2P       | Two-Point Field Goals made per game |
| 2PA      | Two-Point Field Goal attempts per game |
| 2P%      | Two-Point Field Goal percentage (2P ÷ 2PA) |
| FT       | Free Throws made per game |
| FTA      | Free Throw attempts per game |
| FT%      | Free Throw percentage (FT ÷ FTA) |
| ORB      | Offensive Rebounds per game |
| DRB      | Defensive Rebounds per game |
| TRB      | Total Rebounds per game |
| AST      | Assists per game |
| STL      | Steals per game |
| BLK      | Blocks per game |
| TOV      | Turnovers per game |
| PF       | Personal Fouls per game |
| PTS      | Points scored per game |
| CHAMPION | Did that team win the championship that year? 1 = Yes, 0 = No |
| PLAYOFFS | Did that team make the playoffs that year? 1 = Yes, 0 = No |

</details>

---

# Learning Goals

- Learn about the basics of categorical modeling
- Compare and contrast models using statistical metrics
- Learn about Quadratic Discriminant Analysis
- Learn about Linear Discriminant Analysis
- Learn about Logistic Regression
- Learn about K Nearest Neighbors and how it works.

---

# Getting Started


In this module, we will learn about categorical modeling in order to predict the outcome of a certain event based on past and current metrics. 

Today, we will be predicting what teams are going to make the WNBA playoffs this year (2025)! Before we get started, we need to import some Python libraries in order to achieve this goal.

If you don't have any of these librarys downloaded, you can learn about how to do so with the following links:

Pandas: https://pandas.pydata.org/docs/getting_started/install.html 

Numpy: https://numpy.org/install/

Sci-Kit Learn: https://scikit-learn.org/stable/install.html

After downloading those librarys, let's import them.

In [1]:
# Import the necessary librarys for the module
# Basic Data Science Library for importing data and data manipulation
import pandas as pd

# Library for mathematical operations
import numpy as np

# Import only the necessary functions from the sklearn library
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler

---
# Explaination of Statisitics

Before we begin the technical aspect of this learning module, let's go over some key statistics and what they mean. 

_Field Goal_: Any type of basket made during live basketball play. Any type of basket made outside of free throws is considered a field goal. Dunks, layups, mid-range shots and three-pointers are all examples of a field goal.

_2-point shot_: A 2 point shot is any type of field goal that counts for 2 points. These shots are made from inside the 3-point line, which is the the large line painted around a basketball net on one side of a court. 2-point shots includes dunks, layups and midrange shots.

_3-point shot_: A 3 point shot is the reverse of a 2-point shot. It is any shot made outside of the 3-point line. If a player steps over the line during the shot, it will count as a 2-point shot. 

_Rebound_: A rebound is when a player shoots the ball, misses the shot and the same or different player catches the ball after it falls down from the rim. The player who made the rebound must maintain full control of the ball and land on the ground with possesion. 

_Assist_: An assist is made when one player passes the ball to their teammate and that teammate makes a field goal. However, the pass must directly contribute to the field goal that was made. For example, the general rules state that an assist only counts if the teammate who recieved the ball via a pass, makes the field goal in 2 dribbles or less. 

_Free throw_: A free throw is a free, unguarded shot (or 2) that was award to a player via a foul commited by the other team. Usually, free throws are awarded off of shooting fouls which occur when a player attempts to score a basket but a defender on the opposing team makes illegal contact, such as a slap on the wrist while a player is making a shot attempt. 

_Block_: A block is made when a defender swipes an offensive players shot out of the air without making illegal contact with the player who attempted the shot. 

_Steal_: A steal occurs when a defensive player cleanly swipes the ball out of an offensive players hands or when a defensive player intercepts a ball off a pass between two offensive players. 

_Turnover_: A turnover occurs when there is a steal by the defense, a travel violation by the offense (more then two steps without dribbling the ball), an offensive player stepping out of bounds, etc.

---

# Classfication

Classification is a modeling approach to predict a non-numeric (or categorical) outcome using available data. For this module, we will be predicting a binary outcome, like 'yes' or 'no', 'Made Playoffs' or 'Did not make Playoffs', etc. There are various models that can be used to achieve this but today we will go over Logistic Regression, k Nearest Neighbors (kNN), Linear Discriminant Analysis (LDA) and Quadratic Discriminant Aanlaysis (QDA) models. Each of these models use different methodologies to achieve the same goal. Therefore, each models predictions may differ from each other, warranting comparison. This can be done using metrics that we will go over later in this lesson.

---

# Modeling Process For This Module

In order to build a model, we will need to train the model and then test the model to see how accurate it is. Normally, we would want to build the model by splitting a season's worth of data into a train dataset and a test dataset. Generally, you want to split the data into a higher proportion of training data than test data. This ensures that the model will be more adequately prepared to make predicitions on the test data. On the flip side, there needs to be a balance between your train and test sets because the more test data you have, the more data you have to test your models accuracy. Normally, we use 60-80% of the data for training the model and 20-40% for testing the model.

However, we will want to do something a little different in this module. For the seasons we will be training the model with, there are only 12 WNBA teams. That is not enough data! In this case, we have compiled a dataset of 'Per Game' data for every team over the last 5 seasons, before the 2024 season. 'Per Game' means that the stats are an average of each catgory among all games played. For example, the 'PTS' column is an average of each teams points scored per game. We will train our models on that compiled dataset and then test the accuracy of our models with the 2024 season data. After reviewing each model's accuracy score, we will pick the model with highest accuracy and re-train it with another compiled dataset. This other compiled dataset will include 2020-2024 data instead of 2019-2023 in order to provide the newest data to the model. After retraining the model, we will have the model predict which teams will make the WNBA playoffs in the current 2025 season.

NOTE: The compiled dataset was built upon several 'Per Game' datasets from Basketball-Reference.com. Please visit their website using the link at the top of this module. Each teams name has been edited to identify what year each team's stats are from since there are reoccuring team names. A 'CHAMPION' column was also added in order to identify which teams won a championship. 1 = Champions, 0 = Not Champions. We will not be using this column in this lesson. Instead, we will use the 'PLAYOFFS' column since we are predicting who will make the playoffs. The 'PLAYOFFS' column was added after the fact, like the 'CHAMPION' column, and was not in the original Basketball-Reference.com dataset.

In order to test the accuracy of the model(s) amongst one another, we want to train the model(s) on the previous 5 season's worth of data for higher model accuracy. We will train the model(s) using 2019-2023 season data and test it with the 2024 season data. The model with the highest accuracy, or of our choosing, will be used to predict who will make the WNBA playoffs this season using the 'Per Game' stats thus far.

To summarize, a model is something we can train with data so it can make a prediction on new data.

# Importing data and making train/test sets

Let's get started by reading in the data.

In [2]:
# Read in the WNBA 2019-2023 data from Github
# This dataset will be used to train the intial models
TRAIN_WNBA_19_23_Data = pd.read_csv('https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes/WNBA_PER_GAME_19_23_DATA.csv')

In [3]:
# Read in the WNBA 2024 data from Github
# This will be used to test the intial models 
TEST_WNBA_2024_Data = pd.read_csv('https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes/2024_WNBA_Per_Game.csv')

In [4]:
# Read in the WNBA 2020-2024 data from Github 
# This dataset will be used to train the final model
TRAIN_WNBA_20_24_Data = pd.read_csv('https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes/WNBA_PER_GAME_20_24_DATA.csv')

In [5]:
# Read in the current WNBA 2025 stats for our model prediction later on in this module
CURRENT_WNBA_2025_Data = pd.read_csv('https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20For%20Intro%20to%20Categorical%20Model-Making%20Module%20by%20Austin%20Hayes/2025_WNBA_Per_Game.csv')

We need to scale the data in order to even the playing field among variables for model-making. This prevents a model from becoming too biased towards features or variables with larger numerical ranges. Scaling essentially standardizes the data so that all data points are rescaled so that their means are 0 and the standard deviation is 1. Although the data is rescaled, their data ranges are not distorted. The formula is the following: (x - mean) / Standard Deviation. 

Note on fit_transform vs ()transform().

fit_transform() scales the data as detailed above while also calculating the unique parameters of the data. These parameters will be used to scale the test set, which creates new and unique results for the model to make predictions on. This creates less biased and mroe well-rounded results. 

transform() this scales the data as detailed above, but without the fit portion of the function above. This data will be rescaled using the training set's parameters. 

We will use fit_transform() on the training data and transform() on the testing data.

In [6]:
# Make the scaler object
scaler = StandardScaler()

# Fit the scaler object to the training data for intital model-making
training_data_19_23 = scaler.fit_transform(TRAIN_WNBA_19_23_Data[['FG', 'FGA', '3P', '3PA', '2P', '2PA', 'FT', 'FTA', 'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']])


# Scale the test data for intial model-making
test_data_2024 = scaler.transform(TEST_WNBA_2024_Data[['FG', 'FGA', '3P', '3PA', '2P', '2PA', 'FT', 'FTA', 'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']])


# Do the same for the other training data for final model-making
# final training data
training_data_20_24 = scaler.fit_transform(TRAIN_WNBA_20_24_Data[['FG', 'FGA', '3P', '3PA', '2P', '2PA', 'FT', 'FTA', 'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']])

# Scale the current 2025 WNBA data for final model-making
test_data_2025 = scaler.transform(CURRENT_WNBA_2025_Data[['FG', 'FGA', '3P', '3PA', '2P', '2PA', 'FT', 'FTA', 'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']])

Let's get to work now! Remember, you only need to scale your predictors, not your target/variable that you are making a prediction on.

---

# Accuracy Score 

Throughout this module we will be using an accuracy score to access model prediction accuracy. An accuracy score is a metric that essentially divides the total number of correct predictions by the total amount of predictions. 

                                Total number of correct predictions by the model

                                ________________________________________________

                                Total number of predictions made by the model

---

# Logistic Regression

Logistic Regression is a categorical model that uses probabilities to predict a binary outcome. A Logistic Regression model outputs log odds (logit) which range from 0 to 1. 0 means a negative/ false outcome and a 1 is a positive/ true outcome. These odds are calulcated by the following equation, p / (1-p), where p is the probability of an outcome. This essentially assesses an event's probability of success to its probability of failure in the form of a ratio. This is why we will name the predictions made by a Logisitic Regression model, "logit_predictions", throughout this module. When a logit outcomes from a Logistic Regression model assume a linear relationship between predictors and the response vairable.


Let's go ahead and establish a logistic model and train it with the compiled 2019-2023 data.

We will train it ONLY using the neccesary stats. The model also needs to be trained using integer values, so let's keep it simple by predicting the champion based on the following stats: FG, FGA, 3P, 3PA, 2P, 2PA, FT, FTA, ORB, DRB, TRB, AST, STL, BLK, TOV, PF, PTS. The definitions of these variables can be found at the top of this module. 

In [7]:
# Create the logistic regression model
logistic_model = LogisticRegression()

# Train the model using the 2023 WNBA data
logistic_model.fit(training_data_19_23, TRAIN_WNBA_19_23_Data['PLAYOFFS'])

LogisticRegression()

In [8]:
# Have the model make predictions using the 2024 WNBA data
logit_predictions = logistic_model.predict(test_data_2024)

# takes a look at the array of predictions
logit_predictions

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1], dtype=int64)

MMHHHMMM... very interesting! The model predicted 10 teams to make the 2024 WNBA PLayoffs! Unfortunately, there are 8 teams that make the playoffs every year! The New York Liberty won the championship that season, which would be the second team on the list. Let's take a look at the probabilities of each team making the playoffs, according to the model's probabilties.

In [9]:
# Have the model show the probabilities of each team making the playoffs
logistic_model.predict_proba(test_data_2024).round(4)

array([[6.000e-04, 9.994e-01],
       [1.000e-04, 9.999e-01],
       [2.530e-02, 9.747e-01],
       [3.862e-01, 6.138e-01],
       [6.100e-03, 9.939e-01],
       [6.000e-04, 9.994e-01],
       [3.590e-02, 9.641e-01],
       [6.850e-02, 9.315e-01],
       [1.207e-01, 8.793e-01],
       [6.274e-01, 3.726e-01],
       [8.349e-01, 1.651e-01],
       [4.923e-01, 5.077e-01]])

Looking at the output above, the probabilities in the left column are for 0, not making the playoffs, and the right column is for 1, making the playoffs. Remember, these probabilities are in a scientific notation so you can indicate where they actually are beyond the decimal point by looking at the number beyond the hyphen. We can see that the model predicted the second team in the list, the New York Liberty, to have the highest probability of making the playoffs, 0.9999, for the 2024 season. For the model to predict a team as a playoff team, the probability needs to be greater than 50%.  The third team in this list is the Indiana Fever and they have a 0.0253 probability of not making the playoffs.

Quick Note about the probabilites above. Each row adds up to one because probabilities are on a scale between zero and one. In other words, an event cannot have a negative percent chance of happening or over 100% chance of happening. An event occurs or it does'nt; the result is binary. Since the results are binary, the two must add up to 100% because one of the possible outcomes will occur. There is an 100% certainty that one of the possbile outcomes will happen, hence their probabilites add up to one.

Logistic regression assumes a linear relationship between the log odds of the probability of making the playoffs and the features we are using to make that prediction.   In this module we will not cover how to assess whether or not that relationship is linear.  

Below, we will input all of the actual/ oberserved values of which teams made the playoffs as well as the predictions made by the model. The accuracy score function will output a ratio of how may correct predictions were made by the model when compared to the actual results. Remember, this prediction and accuracy score is just for the training dataset. 

In [10]:
# Make accuracy score
accuracy_score_log = accuracy_score(TEST_WNBA_2024_Data['PLAYOFFS'], logit_predictions)

In [11]:
# Print the accuracy score
print(f"Accuracy Score: {accuracy_score_log}")

Accuracy Score: 0.8333333333333334


The Logistic Regression model has a 91.6% accuracy for predicting the teams who made the WNBA playoffs during the 2024 season.

---

# Discriminant Analysis


There are two types of Discriminant Analysis Models that we will be going over in this lesson. One is called Linear Discriminant Analysis (LDA) and the other being Quadratic Discriminant Analysis (QDA). These models essentially create a line to classify data points into a category. Discriminant Analysis does this by selecting the category with the highest value from an estimated Normal probability density from among densities fit to each category.  LDA comes from assuming that each category has the same variance and correlation among the features, while QDA allows each group/category to have a different set of variances and correlations.  We will begin with LDA first.

--- 

# Linear Discriminant Analysis


LDA is a linear form of discriminant analysis and it assumes that each group follows a normal distribution and that the variance and correlationfor each classification group is the same. This means that the variance(s) and correlations between the features for the teams that made the playoffs and the teams that did not make the playoffs is the same. Let's get started; we will follow the same process as we did with logistic regression.

In [12]:
# Create the Linear Discriminant Analysis model
lda_model = LinearDiscriminantAnalysis()

# Train the model using the 2023 WNBA data
lda_model.fit(training_data_19_23, TRAIN_WNBA_19_23_Data['PLAYOFFS'])

LinearDiscriminantAnalysis()

In [13]:
# Now let's test the model using the 2024 WNBA data and have it make predictions
lda_model.predict(test_data_2024)

array([1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0], dtype=int64)

Interesting; this LDA model produced worse results than the logistic model since it only predicted 6 teams to make the playoffs. After scaling the predictor variables, this model has predicted that the top 3 teams, in terms of per game stats, will NOT make the playoffs. Although this is an incorrect result, let's analyze some the probabilities to see if some of the teams predicted as non-playoff teams had close to a 50% chance.

In [14]:
# Predict the probabilities of each team making the playoffs
lda_model.predict_proba(test_data_2024).round(4)

array([[9.900e-03, 9.901e-01],
       [7.000e-04, 9.993e-01],
       [1.052e-01, 8.948e-01],
       [5.568e-01, 4.432e-01],
       [1.310e-02, 9.869e-01],
       [6.000e-04, 9.994e-01],
       [4.665e-01, 5.335e-01],
       [8.270e-02, 9.173e-01],
       [2.166e-01, 7.834e-01],
       [7.878e-01, 2.122e-01],
       [6.474e-01, 3.526e-01],
       [5.004e-01, 4.996e-01]])

WOW! The LDA model predicted the 2 teams with the WORST per game stats to make the playoffs. It is certainly possible but seems a bit fishy since none of the top 3 teams were considered as playoff teams by the model. Let's record the accuracy score and note that this may not be a good model to use. 

In [15]:
# Assign LDA Model Prediction from 2024 WNBA Data to a variable
lda_predictions = lda_model.predict(test_data_2024)

# takes a look at the array of predictions
lda_predictions

array([1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0], dtype=int64)

In [16]:
# Get Accuracy Score
accuracy_score_lda = accuracy_score(TEST_WNBA_2024_Data['CHAMPION'], lda_predictions)

# Print the accuracy score
print(f"Accuracy Score: {accuracy_score_lda}")

Accuracy Score: 0.4166666666666667


The LDA model has a 41.7% accuracy for predicting the teams who made the WNBA playoffs during the 2024 season.

Look at that accuracy score! That is extremely bad and this model should NOT be used for this dataset.

Let's move onto QDA!

---

# Quadratic Disciminant Analysis

QDA is very similar to LDA. QDA, like LDA, assumes a normal distribution. Unlike LDA, QDA allows each classification group to have different variances. QDA does NOT assume a linear classification boundary so it is seen as a flexible option in comparison. Now Let's get started.

In [17]:
# Make the Quadratic Discriminant Analysis model
qda_model = QuadraticDiscriminantAnalysis()

# Train the model using the 2023 WNBA data
qda_model.fit(training_data_19_23, TRAIN_WNBA_19_23_Data['PLAYOFFS'])

QuadraticDiscriminantAnalysis()

In [18]:
# Have the model make predictions using the 2024 WNBA data
qda_predictions = qda_model.predict(test_data_2024)

# takes a look at the array of predictions
qda_predictions

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int64)

Very Interesting! This QDA Model every team to make the playoffs in 2024! Let's take a look at the probability breakdown below.

In [19]:
# Predict the probabilities of each team making the playoffs
qda_model.predict_proba(test_data_2024)

array([[2.48981476e-19, 1.00000000e+00],
       [3.56498825e-34, 1.00000000e+00],
       [6.42154640e-10, 9.99999999e-01],
       [8.03255972e-33, 1.00000000e+00],
       [2.80310658e-56, 1.00000000e+00],
       [1.93734207e-40, 1.00000000e+00],
       [7.78290385e-05, 9.99922171e-01],
       [1.39860160e-30, 1.00000000e+00],
       [4.89903764e-18, 1.00000000e+00],
       [2.05337647e-01, 7.94662353e-01],
       [1.04328806e-01, 8.95671194e-01],
       [6.86106838e-16, 1.00000000e+00]])

In [20]:
accuracy_score_qda = accuracy_score(TEST_WNBA_2024_Data['CHAMPION'], qda_predictions)

# Print the accuracy score
print(f"Accuracy Score: {accuracy_score_qda}")

Accuracy Score: 0.08333333333333333


The QDA model has a 8.3% accuracy for predicting the teams who made the WNBA playoffs during the 2024 season.

As you could have guessed, there is an astonishingly low accuracy score. There might be a massive colinearity issue with using this model! It may be best to skip the use of this model!

Let's move onto KNN!

---

# K Nearest Neigbors (KNN)



K Nearest Neighbors is a type of model that decides where to classify a data point based on the distance from previously clssified data points that the model was trained on. It is essentially deciding where to classify a data point based on the closest points or "neighbors". You can control the number of neighbors that a data point will be classified with by defining K. If you make the K number of neighbors too high, it may make the model too generalized and cause it to have low accuracy. Likewise, if you make K too small, it will be too niche and become inaccurate as well. 

Due to KNN being distance based, you will need to scale the predictor variables in order to get the best results. Scaling data points essentially levels the playing field among different measures in order to prevent the model from becoming too biased towards predictor variables with larger measures. For example, 'FGA', or 'Field Goals Attempted Per Game' will always have a larger range of numbers than 'Offensive Rebounds Per Game'

Let's test it out.

IMPORTANT RULE: For good practice, you should ONLY use odd numbers for K. This is done to prevent a tie during the prediction process. For example, if you used K=4 and there happens to be a 2-2 vote among points, it will likely result in an inaccuracte outcome. 

We will need to use standard scaler to scale the data. Let's go ahead and make the KNN model. Let's start off with K = 5 Neighbors.

In [21]:
# Make the model object for KNN with 5 neighbors
KNN_model_5 = KNeighborsClassifier(n_neighbors=5)

# Fit the model using the 2019-2023 WNBA data that was scaled earlier
KNN_model_5.fit(training_data_19_23, TRAIN_WNBA_19_23_Data['PLAYOFFS'])

KNeighborsClassifier()

In [22]:
# Have the model make predictions using the 2024 WNBA data
knn_predictions_5 = KNN_model_5.predict(test_data_2024)

# takes a look at the array of predictions
knn_predictions_5

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1], dtype=int64)

In [23]:
# calculate accuracy score for KNN model with 5 neighbors
accuracy_score_knn_5 = accuracy_score(TEST_WNBA_2024_Data['PLAYOFFS'], knn_predictions_5)

# Print the accuracy score
print(f"Accuracy Score: {accuracy_score_knn_5}")

Accuracy Score: 0.8333333333333334


The KNN model with 5 neighbors has a 83.3% accuracy for predicting the teams who made the WNBA playoffs during the 2024 season.

Interesting! It predicted one to many teams to make the playoffs. However, this model has the same accuracy score as the logistic model. Now, let's do the same thing with a higher number of neigbors! Let's try a higher K of 51!

In [24]:
# Make the model object for KNN model with 51 neighbors
KNN_model_51 = KNeighborsClassifier(n_neighbors=51)

# Fit the model using the 2019-2023 WNBA data that was scaled earlier
KNN_model_51.fit(training_data_19_23, TRAIN_WNBA_19_23_Data['PLAYOFFS'])

KNeighborsClassifier(n_neighbors=51)

In [25]:
# Have the model make predictions using the 2024 WNBA data
knn_predictions_51 = KNN_model_51.predict(test_data_2024)

# takes a look at the array of predictions
knn_predictions_51

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int64)

In [26]:
# calculate accuracy score for KNN model with 51 neighbors
accuracy_score_knn_51 = accuracy_score(TEST_WNBA_2024_Data['PLAYOFFS'], knn_predictions_51)

# Print the accuracy score
print(f"Accuracy Score: {accuracy_score_knn_51}")

Accuracy Score: 0.6666666666666666


The KNN model with 51 neighbors has a 66.7% accuracy for predicting the teams who made the WNBA playoffs during the 2024 season.

Not great! The model used too many points and predicted every team to make the playoffs. Now let's try K = 15

In [27]:
# Make the model object for KNN model with 15 neighbors
KNN_model_15 = KNeighborsClassifier(n_neighbors=15)

# Fit the model using the 2019-2023 WNBA data that was scaled earlier
KNN_model_15.fit(training_data_19_23, TRAIN_WNBA_19_23_Data['PLAYOFFS'])

KNeighborsClassifier(n_neighbors=15)

In [28]:
# Have the model make predictions using the 2024 WNBA data
knn_predictions_15 = KNN_model_15.predict(test_data_2024)

# takes a look at the array of predictions
knn_predictions_15

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int64)

In [29]:
# calculate accuracy score for KNN model with 15 neighbors
accuracy_score_knn_15 = accuracy_score(TEST_WNBA_2024_Data['PLAYOFFS'], knn_predictions_15)

# Print the accuracy score
print(f"Accuracy Score: {accuracy_score_knn_15}")

Accuracy Score: 0.6666666666666666


The KNN model with 5 neighbors has a 66.7% accuracy for predicting the teams who made the WNBA playoffs during the 2024 season.

This model predicted 9 teams to make the playoffs; one too many. Not a great accuracy score either. 

Now, let's try k=13 and k=21 now!

K=13:

In [30]:
# Make the model object for KNN model with 13 neighbors
KNN_model_13 = KNeighborsClassifier(n_neighbors=13)

# Fit the model using the 2019-2023 WNBA data that was scaled earlier
KNN_model_13.fit(training_data_19_23, TRAIN_WNBA_19_23_Data['PLAYOFFS'])

# Have the model make predictions using the 2024 WNBA data
knn_predictions_13 = KNN_model_13.predict(test_data_2024)

# takes a look at the array of predictions
knn_predictions_13

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1], dtype=int64)

In [31]:
# calculate accuracy score for KNN model with 13 neighbors
accuracy_score_knn_13 = accuracy_score(TEST_WNBA_2024_Data['PLAYOFFS'], knn_predictions_13)

# Print the accuracy score
print(f"Accuracy Score: {accuracy_score_knn_13}")

Accuracy Score: 0.75


The KNN model with 13 neighbors has a 75% accuracy for predicting the teams who made the WNBA playoffs during the 2024 season.

K=21:

In [32]:
# Make the model object for KNN model with 21 neighbors
KNN_model_21 = KNeighborsClassifier(n_neighbors=21)

# Fit the model using the 2019-2023 WNBA data that was scaled earlier
KNN_model_21.fit(training_data_19_23, TRAIN_WNBA_19_23_Data['PLAYOFFS'])

# Have the model make predictions using the 2024 WNBA data
knn_predictions_21 = KNN_model_21.predict(test_data_2024)

# takes a look at the array of predictions
knn_predictions_21

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1], dtype=int64)

In [33]:
# calculate accuracy score for KNN model with 21 neighbors
accuracy_score_knn_21 = accuracy_score(TEST_WNBA_2024_Data['PLAYOFFS'], knn_predictions_21)

# Print the accuracy score
print(f"Accuracy Score: {accuracy_score_knn_21}")

Accuracy Score: 0.75


The KNN model with 21 neighbors has a 75% accuracy for predicting the teams who made the WNBA playoffs during the 2024 season.

Although our models have not produced the most ideal results, this is a part of the modeling experience. You learn what works and what does'nt. We will talk about final conclusions after we make our final prediction on the 2025 data.

Since we have a tie for the highest accuracy between the Logistic model and the KNN Model with K=5, let's have both of them make a prediction. Remember, an accuracy acore cannot be calculated since playoff teams have not been decided yet. Let's try to predict which teams will get in to the playoffs!

---

# Final Predictions:

As discussed earlier, we will retrain each model in order to maintain fresh data; only keeping data from 5 seasons before the season we are predicting. We will use a different dataset that removes the 2019 data and inputs the 2024 season data that we have been making predictions on throughout this whole module. 

Let's go! We will start by making the final logistic model

In [34]:
# Make a final Logistic model using the 2020-2024 WNBA data
logistic_model_final = LogisticRegression()

# Fit the model using the 2020-2024 WNBA data that was scaled earlier
logistic_model_final.fit(training_data_20_24, TRAIN_WNBA_20_24_Data['PLAYOFFS'])

# Have the model make predictions using the 2025 WNBA data
logit_predictions_final = logistic_model_final.predict(test_data_2025)

# takes a look at the array of predictions
logit_predictions_final

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0], dtype=int64)

In [35]:
# Key to know what team the list of probabilities is listing
CURRENT_WNBA_2025_Data

Rk                    Team   G     MP    FG   FGA    FG%   3P   3PA  \
0    1          Minnesota Lynx  44  200.6  32.1  68.1  0.472  9.6  25.4   
1    2      Los Angeles Sparks  44  201.7  30.7  67.1  0.457  8.6  25.6   
2    3           Indiana Fever  44  200.6  31.4  68.8  0.456  8.3  24.0   
3    4           Atlanta Dream  44  200.6  30.1  67.8  0.443  9.6  28.4   
4    5        New York Liberty  44  200.0  29.6  65.3  0.453  9.5  26.8   
5    6          Las Vegas Aces  44  200.0  29.7  67.6  0.439  9.1  25.9   
6    7         Phoenix Mercury  44  200.0  29.7  68.6  0.433  9.4  27.7   
7    8           Seattle Storm  44  201.1  31.1  69.2  0.450  7.5  22.1   
8    9            Dallas Wings  44  200.6  30.2  71.4  0.423  6.6  21.6   
9   10  Golden State Valkyries  44  200.6  26.5  65.2  0.407  9.7  29.9   
10  11      Washington Mystics  44  200.6  28.0  63.7  0.439  5.6  17.1   
11  12             Chicago Sky  44  200.0  28.0  66.0  0.425  6.7  20.2   
12  13         Connecticut Sun  44  200.6  27.6  67.1  0.412  6.2  20.4   

      3P%  ...    FT%   ORB   DRB   TRB   AST  STL  BLK   TOV    PF   PTS  
0   0.378  ...  0.760   8.5  25.6  34.2  23.3  8.2  4.8  12.1  16.0  86.1  
1   0.337  ...  0.769   8.0  25.3  33.3  20.7  7.3  3.1  14.1  16.9  85.7  
2   0.346  ...  0.770   8.8  24.6  33.4  20.6  7.7  2.8  12.8  18.6  84.9  
3   0.337  ...  0.786   8.9  27.7  36.6  21.4  6.8  4.5  11.2  15.8  84.4  
4   0.355  ...  0.837   6.8  26.9  33.7  21.8  7.6  4.4  12.9  16.2  84.4  
5   0.351  ...  0.824   8.1  25.4  33.5  19.6  7.5  4.9  11.9  16.6  83.6  
6   0.340  ...  0.784   8.7  26.0  34.7  21.0  7.8  3.7  12.5  17.8  82.8  
7   0.342  ...  0.771   6.9  24.0  30.9  21.3  8.5  5.1  11.6  17.1  82.1  
8   0.304  ...  0.802  10.0  25.3  35.4  20.2  7.3  4.0  12.8  18.9  81.7  
9   0.325  ...  0.815   8.7  26.6  35.3  18.5  6.8  2.9  12.9  17.5  77.7  
10  0.329  ...  0.754   7.7  25.3  33.0  19.2  6.5  3.2  14.2  18.2  77.1  
11  0.330  ...  0.763   9.8  25.7  35.5  19.3  6.0  4.2  15.5  17.8  75.8  
12  0.302  ...  0.793   8.5  23.3  31.9  17.4  7.8  3.7  13.1  19.6  75.8  

[13 rows x 25 columns]

Looks like the logistic model has predicted 9 teams to make the playoffs. Unfortuantely that is one too many once again.

For our final conclusion we need to look at the probabilties of each team making the playoffs in order to exclude one team.

In [36]:
# Calculate the probabilities of each team making the playoffs using the final logistic model
log_final_probs = logistic_model_final.predict_proba(test_data_2025)

# Print the probabilities of each team making the playoffs
print(log_final_probs)

[[4.70967755e-04 9.99529032e-01]
 [6.84795875e-02 9.31520413e-01]
 [1.22011753e-01 8.77988247e-01]
 [3.51319829e-04 9.99648680e-01]
 [1.06315303e-03 9.98936847e-01]
 [9.03132681e-03 9.90968673e-01]
 [1.85958801e-02 9.81404120e-01]
 [3.55871673e-02 9.64412833e-01]
 [4.24819760e-01 5.75180240e-01]
 [1.11892568e-01 8.88107432e-01]
 [7.83904501e-01 2.16095499e-01]
 [8.45344519e-01 1.54655481e-01]
 [9.60283462e-01 3.97165377e-02]]


Among the teams predicted to make the playoffs, the Valkyries have the lowest chance of making it with a 65.5% chance. Let's drop that team out of the playoffs for our conclusion from this model. 

Therefore the teams predicted to make the playoffs are: The Liberty, Lynx, Sparks, Fever, Dream, Mercury, Aces, and the Storm.

Now, let's try a KNN model with K=5 to see if it produces better results.

K=5:

In [37]:
# Make the model object for KNN model with 5 neighbors
Final_KNN_model_5 = KNeighborsClassifier(n_neighbors=5)

# Fit the model using the 2020-2024 WNBA data that was scaled earlier
Final_KNN_model_5.fit(training_data_20_24, TRAIN_WNBA_20_24_Data['PLAYOFFS'])

# Have the model make predictions using the 2025 WNBA data
Final_knn_predictions_5 = Final_KNN_model_5.predict(test_data_2025)

# takes a look at the array of predictions
Final_knn_predictions_5

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0], dtype=int64)

This model did slightly worse and predicted 10 teams to go to the playoffs. Not only that but the model only chose the top 10 teams in terms of per game stats. Most likely, that is where the boundary for the model is located. 

We won't look at the probabilities for this model since it performed worse than the logistic model. We will make our conclusion(s) using the logistic models' results.

After exploring the results across different models, we can see that the logistic model has produced the best results. Let's move onto the final conclusions for this module.

---

# Final Conclusions


Throughout this module, we have seen a lot of things. We have gone over the basics of modeling with KNN, LDA, QDA and Logistic models. We have discussed scaling, accuracy scores, tie-breakers, etc. There is a LOT that comes with the modeling process. There are also lots of other types of models that could have been used to find this outcome such as SVM's, Decision Trees, etc. There are plenty of other alternatives that could be used in place of the models we have learned about today. 

We learned that Discriminant Analysis was NOT a good idea for this dataset and that logistic models, along with KNN models, fit this project much better. This may be due to the scaling along with the variances/relationships among the predictor variables. Exploratory analysis and data validation is extremely important in the model-making process. Preparing the data for usage and discovering what works and what does'nt is a vital piece of the puzzle, as we figured out today. 

Despite the failures using Discriminant Analysis models, our Logistic models and KNN models produced consistant results among their respective predictions. After modeling the results, we will conclude that the teams making the playoffs in 2025 are the Liberty, Lynx, Sparks, Fever, Dream, Mercury, Aces, and the Storm!

---

# Review Questions

Below, you will be asked a series of questions to review the material you have learned throughout this module.

### 1.) How does KNN work and what needs to be done to the data before use of a KNN model?

### 2.) Using the 2020-2024 dataset, create a list of predictor variables that include FG%, 3P%, 2P%, FT%, ORB, AST. Scale the list of predictor variables. Make sure you also scale them for the 2025 data since you will be asked to make a prediction on them later. Create a new column in the ORIGINAL 2020-2024 dataset that assesses whether a teams points ('PTS') are over 82 per game. 1 = True, 0 = False. Make this column for the ORIGINAL 2025 dataset as well. 

Hint: _Use .astype(int)_ 

### 3.) Using the scaled data you created in question 2, predict which team(s) will have above 82 points per game during the 2025 WNBA season with a QDA model. Record an accuracy score. 

QUICK NOTE: You can use an accuracy score in this case because there are already pts per game recorded for each team thus far. Although they are not final for the WHOLE season, this is an interesting experiment to see if these are strong predictors of whether a team will have high points per game. 

### 4.) Do the same thing you did in question 3 but with 3 KNN models instead. Use K=5, K=25 and K=51 respectively. Record an accuracy score.

### 5.) As you did in questions 3 and 4, make a prediction but use a Logistic model instead. Record an accuracy score.

### 6.) Which model(s) had the highest accuracy score? Is there a tie between the most accurate models? Or is there a model that is clearly the best in terms of an accuracy score?

### 7.) What differentiates LDA and QDA? List a few differences.